Tags query language

jotak opened the pull request:
Big PR to review ;)
Implementing the tags query language (see #52 )
Don't merge til corresponding PR on server side is not merged (language could still evolve?)

jotak commented:
It's ready to review / merge

jotak commented:
@tsegismont do you have some time for this?

karelhala commented on src/tagsQLController.js:
```diff
@@ -0,0 +1,395 @@
+    }
+  }
+  return strTags;
+}
```
I think that this is overcomplicated. Can you give json example of `segments` so we can think of something more readable?

jotak replied:
Sure, I'll add the example in code too:
```javascript
// Input segments: ["fruit", "is in", "pear", "apple", "peach", "<plus-button>", "AND", "color", "=", "green", "<plus-button>"]
// Output string: "fruit IN [pear, apple, peach] AND color=green"
```

I'm happy if you find a way to simplify. I try to keep it relatively straightforward.

karelhala commented on src/tagsQLController.js:
```diff
@@ -0,0 +1,395 @@
+
+function valuesToString(segments, i) {
+  var values = "";
+  var sep = "";
```
Use `let instead of `var`

karelhala commented on src/tagsQLController.js:
```diff
@@ -0,0 +1,395 @@
+function valuesToString(segments, i) {
+  var values = "";
+  var sep = "";
+  while (i < segments.length && segments[i].type === 'value') {
```
Might be better to use `take`Right` from lodash and `findIndex` to find which index had type different than `value`

karelhala commented on src/tagsQLController.js:
```diff
@@ -0,0 +1,395 @@
+  };
+}
+
+export function stringToSegments(strTags, segmentFactory) {
```
Same as above, feels overcomplicated, care to share JSON example so we can think of something more readable?

jotak replied:
That's the reverse function, so switching above's input/output:
```javascript
// Input string: "fruit IN [pear, apple, peach] AND color=green"
// Output segment values: ["fruit", "is in", "pear", "apple", "peach", "<plus-button>", "AND", "color", "=", "green", "<plus-button>"]
```

Here again, there's some functional complexity.

karelhala commented on src/tagsQLController.js:
```diff
@@ -0,0 +1,395 @@
+}
+
+function readEnumeration(strTags, cursor) {
+  var values = [];
```
Use `let` instead of `var`

jotak commented on spec/variables_spec.js:
```diff
@@ -47,4 +47,18 @@ describe('Variables', function () {
   });
+
+  it('should resolve to string', function (done) {
+    ctx.templateSrv.replace = function(target, vars) {
```
Hmm strictly speaking you're right but I prefer to keep here the more verbose form, to make it clear we expect "$container" and not anything else, to keep the test easy to understand.

jotak commented on src/datasource.js:
```diff
@@ -25,16 +26,26 @@ export class HawkularDatasource {
+      })
+      .filter(target => target.id !== undefined
+         || (target.tags !== undefined && target.tags.length > 0)
+         || (target.tagsQL !== undefined && target.tagsQL.length > 0));
```
This is something I fear (probably due to ignorance) when I write javascript: if I write `if (target.id)` instead of `if (target.id !== undefined)`, what happens if `target.id` is 0? or the string "0" ?

jotak replied:
(that's why I tend to write the more verbose way)

jotak commented on src/tagsKVPairsController.js:
```diff
@@ -0,0 +1,106 @@
+      return this.getTagKeys();
+    } else if (segment.type === 'key')  {
+      return this.getTagKeys()
+          .then(keys => [angular.copy(this.removeTagsSegment)].concat(keys));
```
jslearnttoday++

jotak commented on src/tagsQLController.js:
```diff
@@ -0,0 +1,395 @@
+    segments.push(this.uiSegmentSrv.newPlusButton());
+    // Fix plus-button: add it at the end of each enumeration
+    var isInEnum = false;
+    for (var i = 0; i < segments.length; i++) {
```
I was not sure since I modify the array during the loop (splice).... That's why I preferred to keep an explicit increment index

jotak commented on src/tagsQLController.js:
```diff
@@ -0,0 +1,395 @@
+      var i = $index - 2;
+      while (segments[i].type !== 'key') {
+        i--;
+      }
```
Here it looks for the key that precedes the current index. 'findIndex' apparently can search from a given index, but incrementally ; I need reverse order here.

jotak commented:
@karelhala thanks for sharing some javascript practices!

I addressed most of the issues, not all :

- string templates: I am not yet fully convinced on that, readability is kinda subjective sometimes, and to me adding yet another way to represent string is just more confusion... well, maybe I need some time to get used to it :)

- I looked at your suggestions with lodash take[while], but here it's maybe not as obvious as you would think and, in the end, it would not really compact the code.

I think apart from that, I've done everything

jotak commented:
@karelhala , some follow-ups after I used your code snippet: after some small changes I did (that you can see here: https://paste.fedoraproject.org/paste/wMBI5aLdpZ2YRHtT~npLZV5M1UNdIGYhyRLivL9gydE=
) due to the fact that I gave you just some sample input, and not an exhaustive list of the possibilities, it happens that there's still some non working case, with the "not exists" operator which, unlike other operators, requires to write the tag key before the operator when outputing to string: `"mytag" "doesn't exists"` translates to `NOT mytag`.

So rather than coding some special behavior in the reduce function, I prefer to stick to my initial approach which is maybe less condensed, but IMO it shows the language rules more explicitly which is an advantage for understanding them.

BTW I added more unit tests since I discovered the "not exist" operator was not covered.

karelhala commented:
@jotak oh I see, no problem the reduce function was not readable as well, so no problem.

jotak commented:
Thanks for the review Karel!
